### PR TITLE
common: Use thread-local pointer variables to save the *shard

### PR DESCRIPTION
**Pull request by leeq2016**
…d by this thread.

Signed-off-by: liqiang <liqiang64@huawei.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by leeq2016** [comment]
jenkins test make check

File: src/common/mempool.cc [hunk 1]
```diff
@@ -92,9 +92,9 @@ size_t mempool::pool_t::allocated_items() const
 
 void mempool::pool_t::adjust_count(ssize_t items, ssize_t bytes)
 {
-  shard_t *shard = pick_a_shard();
-  shard->items += items;
-  shard->bytes += bytes;
+  thread_shard = (thread_shard == NULL) ? pick_a_shard() : thread_shard;
```
Comment (ronen-fr) [inline src/common/mempool.cc:95]: nullptr

File: src/common/mempool.cc [hunk 2]
```diff
@@ -15,7 +15,7 @@
 #include "include/mempool.h"
 #include "include/demangle.h"
 
-
+static thread_local mempool::shard_t *thread_shard = NULL;
```
Comment (ronen-fr) [inline src/common/mempool.cc:18]: nullptr. Maybe {nullptr} instead of '='?

File: src/common/mempool.cc (hunk 1 above)
Comment (leeq2016) [reply src/common/mempool.cc]: Thanks, resolved.

**Comment by leeq2016** [comment]
jenkins test docs

**Comment by leeq2016** [comment]
jenkins test make check
jenkins test docs

**Comment by leeq2016** [comment]
jenkins test make check

**Comment by leeq2016** [comment]
jenkins test docs

**Comment by leeq2016** [comment]
jenkins test docs

**Comment by leeq2016** [comment]
jenkins render docs

**Comment by leeq2016** [comment]
jenkins test make check

**Comment by leeq2016** [comment]
jenkins test docs

**Comment by leeq2016** [comment]
@tchaikov @ronen-fr Please help to review again, thanks!

**Comment by ronen-fr** [comment]
LGTM.

**Comment by amathuria** [comment]
> LGTM. @amathuria - can you take a look b4 approval? tx

LGTM too!

**Review by ronen-fr — APPROVED** [review]

**Comment by leeq2016** [comment]
@tchaikov Can this PR be merged?

**Comment by ljflores** [comment]
@leeq2016 can you fix your commit titles to be prefixed with the component you are editing? In your case: `common: <rest of commit title>`

**Comment by ljflores** [comment]
http://pulpito.front.sepia.ceph.com/?branch=wip-yuri6-testing-2022-02-14-1456

Failures, unrelated:
    https://tracker.ceph.com/issues/52124
    https://tracker.ceph.com/issues/51076
    https://tracker.ceph.com/issues/54438

Details:
    1. Invalid read of size 8 in handle_recovery_delete() - Ceph - RADOS
    2. "wait_for_recovery: failed before timeout expired" during thrashosd test with EC backend. - Ceph - RADOS
    3. test/objectstore/store_test.cc: FAILED ceph_assert(bl_eq(state->contents[noid].data, r2)) in function 'virtual void SyntheticWorkloadState::C_SyntheticOnClone::finish(int)' - Ceph - RADOS

**Pull request merged**